package/MDAnalysis/analysis/pca.py

```diff
@@ -0,0 +1,240 @@
+
+        Parameters
+        ----------
+
+        n_components : int, optional
+            The number of principal components to be saved, default saves
+            all principal components, Default: -1
+        start : int, optional
+            First frame of trajectory to use for generation
+            of covariance matrix, Default: 0
+        stop : int, optional
+            Last frame of trajectory to use for generation
+            of covariance matrix, Default: -1
+        step : int, optional
+            Step between frames of trajectory to use for generation
+            of covariance matrix, Default: 1
+
+        Return
+        ------
+        cumulated_variance: array, (n_components, )
+            The amount of variance explained by the nth component and the n-1
+            components preceding it.
+        p_components: array, (n_components, n_atoms * 3)
+
+        """
+        self._setup_frames(self._u.trajectory, start, stop, step)
+        self.start = start
+        self.stop = stop
+        self.step = step
+
+        self.n_components = n_components
+        self.run()
+        self._calculated = True
+
+        if n_components is None:
+            return self.cumulated_variance, self.p_components
+        else:
+            return (self.cumulated_variance[:n_components],
+                    self.p_components[:n_components])
+
+    def _prepare(self):
+        n_dim = self._n_atoms * 3
+        self.cov = np.zeros((n_dim, n_dim))
+        self.mean = np.zeros(n_dim,)
+        for i, ts in enumerate(self._u.trajectory[self.start:self.stop:self.step]):
+            x = self._atoms.positions.ravel()
+            self.mean += x
+        self.mean /= self.n_frames
+
+    def _single_frame(self):
+        x = self._atoms.positions.ravel()
+        x -= self.mean
+        self.cov += np.dot(x[:, np.newaxis], x[:, np.newaxis].T)
+
+    def _conclude(self):
+        self.cov /= self.n_frames - 1
+        e_vals, e_vects = np.linalg.eig(self.cov)
+        sort_idx = np.argsort(e_vals)[::-1]
+        self.variance = e_vals[sort_idx]
+        self.p_components = e_vects[sort_idx]
+        self.cumulated_variance = (np.cumsum(self.variance) /
+                                   np.sum(self.variance))
+
+    def transform(self, atomgroup=None, n_components=None):
+        """Apply the dimensionality reduction on a trajectory
+
+        Parameters
+        ----------
+        n_components: int, optional
+            The number of components to be projected onto, default maps onto
+            all components.
+        Returns
+        -------
+        pca_space : array, shape (number of frames, number of components)
+        """
+        if atomgroup is not None:
+            atoms = atomgroup
+        else:
+            atoms = self._atoms
+
+        if self._atoms != atoms:
+            warnings.warn('This is a transform for different atom types.')
+
+        xyz = np.array([atoms.positions.copy() for ts in self._u.trajectory])
```

Comment:
again loading potentially a lot of data into memory. This needs to be in a loop you write yourself instead of using `np.dot`.